Adding swagger updates to support the new NRP resource "DSCP Configuration"

davelasc opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Service team MUST add the "**WaitForARMFeedback**" label if the management plane API changes fall into one of the below categories. 
  - adding/removing APIs.
  - adding/removing properties.
  - adding/removing API-version. 
  - adding a new service in Azure.

<i>Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.</i>

- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://aka.ms/SwaggerPRReview).

davelasc committed:
Adding the swagger for the new resource DSCP Configuration. Also updated the Network Interface to return it with its get function

davelasc committed:
Updated the definition to the updated version in the source code

davelasc committed:
Merge branch 'network-june-release' into users/davelasc/dscp-configuration-update

davelasc committed:
Fix some errors being thrown by the validator (regarding int/ushort typing). Updated the readme.md to include the dscpConfiguration file

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Fixed the erroneous reference within the dscpConfiguration.json to the NetworkInterface definition

AutorestCI commented:
/azp run automation - sdk

MyronFanQiu commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/dscpConfiguration.json:
```diff
@@ -0,0 +1,344 @@
+            }
+          },
+          "x-ms-long-running-operation": true,
+          "x-ms-long-running-operation-options": {
```
Is get method a long running operation as well? It's a little bit rare and just want to confirm.

davelasc replied:
Removing it. Double checked with how the networkInterface.json handles its getters.

MyronFanQiu commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/examples/DscpConfigurationCreate.json:
```diff
@@ -0,0 +1,156 @@
+          "qosTrafficProfiler": {
+            "properties": {
+              "markings": [46],
+              "sourceIpRanges": [
```
Since all of these properties are list object, could we use more than one item in the examples?

davelasc replied:
Updated accordingly

davelasc replied:
Not sure why it is not showing as outdated.

davelasc committed:
Removed the long running operation flag.
Updated the examples to return two markings

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Added more objets within the lists.

AutorestCI commented:
/azp run automation - sdk

MyronFanQiu commented:
@davelasc Could you fix CI errors according to the instruction and error mesages?

davelasc committed:
Removed the duplicated key

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
The type 'ushort ' is not accepted. Turn it into an integer. Hopefully this fixes the issue regarding the QosPortRange properties

AutorestCI commented:
/azp run automation - sdk

davelasc commented:
@MyronFanQiu having issues understanding why the readme is throwing the duplicate exception. Trying to add the directive to suppress warnings about the missing name, id and type properties since they will be inherited. By any chance you know why this is cause or know someone that I can talk to about this.

Right now that change has been removed to (hopefully) have fixed all other issues

MyronFanQiu commented:
@davelasc I checked the error message and I think they gave enough instructions. Each CI error implies that something wrong in your swagger spec like example, conflict or network validation. Usually we cannot suppress these CI errors. If you have specific question for some of them, please let me know and I can invite corresponding authors.

davelasc committed:
Added the custome words dscp and qos in order to fix the error of the words being unkown. Refactor the allocatedNetworkInterfaces to associatedNetworkInterfaces to match the examples with the specification. Removed the QosTrafficProfiler from the examples as it no longer exists. Made sure the delete is using dscpName rather than nicName. Added the period at the end of the description.

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
One traffic profiler slipped through the cracks, removed it. Properly formatted the object for the Qos Port/IP ranges. Was placing a string rather than an integer in the get example.

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Well, I was being silly here. Had a typo hahahahaha

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Added the warning suppression to the readme file. Run the prettier on the specification

AutorestCI commented:
/azp run automation - sdk

davelasc commented:
@MyronFanQiu been able to fix most of the issues, some are still confusing me. The number one is regarding the readme file. Why is it saying that it is duplicated? Need that entry to fix the error of the id, name and type are not being define in the dscpConfiguration.json. Those 3 are inherited from the resource.json.

davelasc committed:
Thought I only had to run the prettier on the specification. Did the same for the examples and the NetworkInterfaceGet.json.

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Re-run the command with the full command placed in the documentation. Hopefully this fixes the prettier errors

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Corrected the typo

AutorestCI commented:
/azp run automation - sdk

akning-ms commented on specification/network/resource-manager/readme.md:
```diff
@@ -1601,6 +1602,9 @@ directive:
   - suppress: RequiredPropertiesMissingInResourceModel
     from: ipAllocation.json
     reason: name, id and type properties are inherited from the upper level
+    suppress: RequiredPropertiesMissingInResourceModel
```
Here is reason for error duplicate key. you should put - suppress:    "-" is missed.
But I noticed there is huge suppression for lint error which we may need to review whether it is still need. as some of them maybe temporary

davelasc replied:
By now this should be fixed. I was just being silly here.

davelasc committed:
Added the proper formatting for the suppression.

AutorestCI commented:
/azp run automation - sdk

davelasc committed:
Added the integer format

AutorestCI commented:
/azp run automation - sdk

davelasc commented:
@MikhailTryakhov @number213 @MyronFanQiu @v-shhatt just a friendly reminder to review the PR. Hopefully it can get approved :D

majastrz commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/dscpConfiguration.json:
```diff
@@ -0,0 +1,343 @@
+  "info": {
+    "title": "NetworkManagementClient",
+    "description": "The Microsoft Azure Network management API provides a RESTful set of web services that interact with Microsoft Azure Networks service to manage your network resources. The API has entities that capture the relationship between an end user and the Microsoft Azure Networks service.",
+    "version": "2020-06-01"
```
>2020-06-01 [](start = 16, length = 10)

Has the `2020-06-01` been released to customers?

davelasc replied:
By the end of this month it should be supported, as far as I'm aware. Did notice yesterday while doing a little bit of support that NRP is already supporting in production this version.

majastrz commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/dscpConfiguration.json:
```diff
@@ -0,0 +1,343 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/dscpConfiguration/{dscpConfigurationName}": {
```
>dscpConfiguration [](start = 100, length = 17)

ARM resource type names should be plural. Can you rename to `dscpConfigurations`?

davelasc replied:
Refactor accordingly

majastrz commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/dscpConfiguration.json:
```diff
@@ -0,0 +1,343 @@
+      }
+    }
+  },
+  "paths": {
```
>paths [](start = 3, length = 5)

Is this going to be a tracked or proxy resource?

If tracked you will also need to support these list calls:
- GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/dscpConfigurations
- GET /subscriptions/{subscriptionId}/providers/Microsoft.Network/dscpConfigurations

If proxy, you will need at least this list call:
- GET /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/dscpConfigurations

davelasc replied:
This is going to be a tracked resource. Adding the pending calls and examples accordingly

majastrz commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/networkInterface.json:
```diff
@@ -1139,6 +1139,10 @@
           "description": "A list of references to linked BareMetal resources."
         },
+        "dscpConfiguration": {
+          "$ref": "./dscpConfiguration.json#/definitions/DscpConfiguration",
```
>DscpConfiguration [](start = 57, length = 17)

This will cause the payload of the entire dscpConfiguration resource to be included in the payload of the network interface resource. This is basically an anti-pattern for the following reasons:
- breaks RBAC because a user who has /read permission on a nic but not a dscp configuration should not be able to see the the dscp configuration
- break Policy an unamiguous alias for properties of dscpConfiguration does not exist
- breaks Azure Resource Graph and Change Tracking because a single change notification represents an update of 2 entities

The alternative is to have the `dscpConfiguration` property be a string that points to a resource ID of the dscpConfiguration resource or have it be an object with an id property. (NRP uses the latter pattern if I recall correctly in some places.)

majastrz replied:
On second thought, it appears that your example is following the recommended pattern, but the schema definition implies the whole dscp configuration payload. Can you fix the schema to match the example? (NRP should have a common type that you can reuse for this already.)

---
In reply to: [451222161](https://github.com/Azure/azure-rest-api-specs/pull/9954#discussion_r451222161) [](ancestors = 451222161)

davelasc replied:
Did the change so it is now using the sub resource reference rather than the full reference of it.

majastrz reviewed (changes requested):
I added a few comments and questions. Please take a look.

davelasc committed:
Added the api calls for list and list all.
Made sure the name is plural
Added the examples for the new api calls
In Network Interface, now it is using the ID of the DSCP configuration rather than the full payload

AutorestCI commented:
/azp run automation - sdk

davelasc reviewed (commented):
Updated the files, please let me know if I successfully address the comments or what needs a little of more polish

davelasc committed:
Forgot to switch the operation id from network interface to dscp configuration

AutorestCI commented:
/azp run automation - sdk

majastrz commented on specification/network/resource-manager/Microsoft.Network/stable/2020-06-01/dscpConfiguration.json:
```diff
@@ -0,0 +1,442 @@
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/dscpConfigurations": {
+      "get": {
```
>get [](start = 7, length = 3)

You have paging declared on list by Sub but not on this operation. Can you add it here too?

davelasc replied:
Added the missing link definition

majastrz reviewed (approved):
Signed off from ARM side. Please add the missing paging on "list by RG", though.

davelasc committed:
Added the missing next link for the list by resource

AutorestCI commented:
/azp run automation - sdk

mmyyrroonn committed:
Merge branch 'network-june-release' into users/davelasc/dscp-configuration-update

AutorestCI commented:
/azp run automation - sdk
